File: src/main/java/de/bixilon/minosoft/data/accounts/MojangAccount.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.accounts;

import com.google.gson.JsonObject;
import de.bixilon.minosoft.Minosoft;
import de.bixilon.minosoft.util.Util;
import de.bixilon.minosoft.util.mojang.api.MojangAuthentication;
import de.bixilon.minosoft.util.mojang.api.exceptions.AuthenticationException;
import de.bixilon.minosoft.util.mojang.api.exceptions.MojangJoinServerErrorException;
import de.bixilon.minosoft.util.mojang.api.exceptions.NoNetworkConnectionException;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class MojangAccount extends Account {
    protected final String id;
    protected final String email;
    protected String accessToken;
    protected RefreshStates lastRefreshStatus;
    protected boolean needsRefresh = true;

    public MojangAccount(String username, JsonObject json) {
        super(json.getAsJsonObject("selectedProfile").get("name").getAsString(), Util.getUUIDFromString(json.getAsJsonObject("selectedProfile").get("id").getAsString()));
        this.accessToken = json.get("accessToken").getAsString();
        this.id = json.getAsJsonObject("user").get("id").getAsString();
        this.email = username;
    }

    public MojangAccount(String accessToken, String id, UUID uuid, String username, String email) {
        super(username, uuid);
        this.accessToken = accessToken;
        this.id = id;
        this.email = email;
    }

    public static MojangAccount deserialize(Map<String, Object> json) {
        return new MojangAccount((String) json.get("accessToken"), (String) json.get("id"), Util.getUUIDFromString((String) json.get("uuid")), (String) json.get("username"), (String) json.get("email"));
    }

    public Map<String, Object> serialize() {
        Map<String, Object> json = new HashMap<>();
        json.put("id", this.id);
        json.put("accessToken", this.accessToken);
        json.put("uuid", getUUID().toString());
        json.put("username", getUsername());
        json.put("email", this.email);
        json.put("type", "mojang");
        return json;
    }

    public void join(String serverId) throws MojangJoinServerErrorException, NoNetworkConnectionException {
        MojangAuthentication.joinServer(this, serverId);
    }

    @Override
    public boolean select() {
        if (this.needsRefresh) {
            return refreshToken() != RefreshStates.ERROR;
        }
        return true;
    }

    @Override
    public void logout() {
        Minosoft.getConfig().getConfig().getAccount().getEntries().remove(this.getId());
        Minosoft.getConfig().saveToFile();
    }

    @Override
    public String getId() {
        return this.id;
    }

    public RefreshStates refreshToken() {
        try {
            this.accessToken = MojangAuthentication.refresh(this.accessToken);
            this.lastRefreshStatus = RefreshStates.SUCCESSFUL;
        } catch (NoNetworkConnectionException e) {
            e.printStackTrace();
            this.lastRefreshStatus = RefreshStates.FAILED;
        } catch (AuthenticationException e) {
            e.printStackTrace();
            this.lastRefreshStatus = RefreshStates.ERROR;
        }
        return this.lastRefreshStatus;
    }

    public String getAccessToken() {
        return this.accessToken;
    }

    public void setAccessToken(String accessToken) {
        this.accessToken = accessToken;
        this.needsRefresh = false;
    }

    public String getEmail() {
        return this.email;
    }


    public boolean needsRefresh() {
        return this.needsRefresh;
    }

    public void setNeedRefresh(boolean needsRefresh) {
        this.needsRefresh = needsRefresh;
    }

    public enum RefreshStates {
        SUCCESSFUL,
        ERROR, // account not valid anymore
        FAILED // error occurred while checking -> Unknown state
    }
}

File: src/main/java/de/bixilon/minosoft/data/accounts/types/MojangAccount.kt
/*
 * Minosoft
 * Copyright (C) 2021 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.accounts.types

import com.fasterxml.jackson.annotation.JsonProperty
import de.bixilon.kutil.cast.CastUtil.nullCast
import de.bixilon.kutil.cast.CastUtil.unsafeCast
import de.bixilon.minosoft.data.accounts.Account
import de.bixilon.minosoft.data.player.properties.PlayerProperties
import de.bixilon.minosoft.data.registries.CompanionResourceLocation
import de.bixilon.minosoft.data.registries.ResourceLocation
import de.bixilon.minosoft.util.KUtil.asUUID
import de.bixilon.minosoft.util.KUtil.toResourceLocation
import de.bixilon.minosoft.util.account.AccountUtil
import de.bixilon.minosoft.util.http.HTTP2.postJson
import de.bixilon.minosoft.util.http.exceptions.AuthenticationException
import de.bixilon.minosoft.util.logging.Log
import de.bixilon.minosoft.util.logging.LogLevels
import de.bixilon.minosoft.util.logging.LogMessageType
import de.bixilon.minosoft.util.nbt.tag.NBTUtil.asCompound
import java.util.*

@Deprecated("Mojang authentication is legacy. Will be removed in the future!")
class MojangAccount(
    override val id: String,
    username: String,
    val uuid: UUID,
    val email: String,
    @field:JsonProperty private var accessToken: String,
    override val properties: PlayerProperties?,
) : Account(username) {
    @Transient
    private var refreshed: Boolean = false
    override val type: ResourceLocation = RESOURCE_LOCATION

    override fun join(serverId: String) {
        AccountUtil.joinMojangServer(username, accessToken, uuid, serverId)
    }

    override fun logout(clientToken: String) {
        val response = mutableMapOf(
            "accessToken" to accessToken,
            "clientToken" to clientToken,
        ).postJson(MOJANG_URL_INVALIDATE)


        if (response.statusCode != 200) {
            throw AuthenticationException(response.statusCode)
        }

        Log.log(LogMessageType.AUTHENTICATION, LogLevels.VERBOSE) { "Mojang account login successful (username=$username)" }
    }

    override fun verify(clientToken: String) {
        if (refreshed) {
            return
        }
        refresh(clientToken)
    }

    fun refresh(clientToken: String) {
        val response = mutableMapOf(
            "accessToken" to accessToken,
            "clientToken" to clientToken,
        ).postJson(MOJANG_URL_REFRESH)

        response.body!!

        if (response.statusCode != 200) {
            throw AuthenticationException(response.statusCode, response.body["errorMessage"].nullCast())
        }

        this.accessToken = response.body["accessToken"].unsafeCast()

        refreshed = true
        Log.log(LogMessageType.AUTHENTICATION, LogLevels.VERBOSE) { "Mojang account refresh successful (username=$username)" }
    }

    override fun toString(): String {
        return "MojangAccount{$username}"
    }

    companion object : CompanionResourceLocation {
        private const val MOJANG_URL_LOGIN = "https://authserver.mojang.com/authenticate"
        private const val MOJANG_URL_REFRESH = "https://authserver.mojang.com/refresh"
        private const val MOJANG_URL_INVALIDATE = "https://authserver.mojang.com/invalidate"
        override val RESOURCE_LOCATION: ResourceLocation = "minosoft:mojang_account".toResourceLocation()

        fun login(clientToken: String, email: String, password: String): MojangAccount {
            val response = mutableMapOf(
                "agent" to mutableMapOf(
                    "name" to "Minecraft",
                    "version" to 1,
                ),
                "username" to email,
                "password" to password,
                "clientToken" to clientToken,
                "requestUser" to true,
            ).postJson(MOJANG_URL_LOGIN)

            response.body!!

            if (response.statusCode != 200) {
                throw AuthenticationException(response.statusCode, response.body["errorMessage"]?.nullCast())
            }

            Log.log(LogMessageType.AUTHENTICATION, LogLevels.VERBOSE) { "Mojang login successful (email=$email)" }

            val uuid = response.body["selectedProfile"].asCompound()["id"].toString().asUUID()
            return MojangAccount(
                id = response.body["user"].asCompound()["id"].unsafeCast(),
                username = response.body["selectedProfile"].asCompound()["name"].unsafeCast(),
                uuid = uuid,
                email = email,
                accessToken = response.body["accessToken"].unsafeCast(),
                properties = PlayerProperties.fetch(uuid),
            )
        }
    }
}
